src/gcp/cloudfunctions.ts

```diff
@@ -260,6 +253,73 @@ export async function setIamPolicy(options: IamOptions): Promise<void> {
   }
 }
 
+/**
+ * Formats the service account to be used with IAM API calls, a vaild service account string is
+ * 'service-account@project.iam.gserviceaccount.com', 'service-account@', or 'service-account'.
+ * @param serviceAccount the custom service account created by the user
+ * @param projectId the ID of the current project
+ * @returns a correctly formatted service account string
+ *
+ * @throws {@link Error} if the supplied service account string is empty
+ */
+function formatServiceAccount(serviceAccount: string, projectId: string): string {
+  if (serviceAccount.length === 0) {
+    throw new Error("Service account cannot be an empty string");
+  }
+
+  const suffix = `@${projectId}.iam.gserviceaccount.com`;
+  if (
+    serviceAccount.length > suffix.length &&
+    serviceAccount.slice(serviceAccount.length - suffix.length) === suffix
+  ) {
+    return `serviceAccount:${serviceAccount}`;
+  }
+
+  const emailId =
+    serviceAccount.charAt(serviceAccount.length - 1) === "@"
+      ? `${serviceAccount}${projectId}.iam.gserviceaccount.com`
+      : `${serviceAccount}@${projectId}.iam.gserviceaccount.com`;
+
+  return `serviceAccount:${emailId}`;
+}
+
+/**
+ * Generates the correct IAM Policy for changing the invoker for HTTP triggered functions.
+ * @param projectId the ID of the current project
+ * @param invoker the member(s) that will be granted the invoker role
+ * @returns correctly formatted IAM policy for use in IAM API calls
+ */
+export function generateIamPolicy(
+  projectId: string,
+  invoker: backend.Invoker | backend.Invoker[] = "public"
+): iam.Policy {
```

Comment:
I would probably take `invoker: backend.Invoker[]` always, special case if any of the elements is "private" first, if any of them are "public" next, and then fall back to `formatServiceAccount`

Reply:
Updated signature with the function rewrite